### ISSUE TITLE
Flatten Schemas

### ISSUE THREAD

**Issue by 17twenty** [golang-migrate/migrate#438]
**Is your feature request related to a problem? Please describe.**
We now have over 40 migrations applied to our database. The issue is, it makes it hard to understand the final schema state by reading the migrations.

**Describe the solution you'd like**

I would love to be able to roll up all exisiting migrations up until the schema_migration point and dump them to a single flattened schema file.

**Describe alternatives you've considered**
Currently our devs can only use the existing database and running a copy of the migrations to get to end state but it makes triggers/enum changes and updates hard to follow still 

**Additional context**

**Comment by 17twenty** [comment]
I'm happy to contribute to this btw, assuming we can get a desired path for it and validation as an issue

**Comment by antonklava** [comment]
I can imagine it being a bit tricky to accomplish this with all the different database drivers.

If you are using MySQL or PostgreSQL maybe something like this can help in the meantime?

```bash
mysqldump --no-data <database>
# or
pg_dump --schema-only <database>
```

**Comment by dhui** [comment]
@17twenty 
If your goal is to view the final output schema, migrate can't do this for you as it treats all migrations as opaque blobs. You can view the final output schema by running all migrations on an empty DB and dump the subsequent schema, as @antonklava has suggested.

If your goal is to reduce migration file clutter by flatten all migrations to a single migration, I see a few potential solutions:
1. Concatenate all migrations to a single migration, replacing the old migrations. Note: not all DB drivers support running multiple statements in a single migration.
1. Run all migrations on an empty DB and save the subsequent schema as a new migration, replacing the old migrations. This may not be exactly the same as the original migrations since there may be other tables/config/params etc in the dumped schema

If you can think of other solutions, I'm all ears. Also note, since `migrate` [doesn't detect conflicting migrations](https://github.com/golang-migrate/migrate/issues/179#issuecomment-475821264), you'll "lock" migration development when flattening, since any migration added by developer B while developer A is flattening migrations will be lost.

How would `migrate` improve either of these processes?

**Comment by 17twenty** [comment]
Great questions guys and thanks for the followup. 

I guess the value add envisioned is for `migrate` to abstract the difference between the database schema dump commands in a uniform way and extract the last version into `20200909-CompactedMigration.up.sql` which would prevent dirty flag issues which we run risk of when doing this manually.
 
The things wanted for a v1 of this would be enums, triggers and table schemas -- does that make sense? It's the sort of things that would be massively beneficial - i believe flydata/liquibase offer similar for the Java community.

**Comment by dhui** [comment]
> I guess the value add envisioned is for migrate to abstract the difference between the database schema dump commands in a uniform way and extract the last version

Can you elaborate? Remember, `migrate` treats migrations as opaque blobs, so parsing or interpreting the migration is out of scope. `migrate` is not an ORM and does not have its own schema so it cannot examine a DB and "diff" it to create a migration. If you're looking for a package that treats the DB as the source of truth, checkout [sqlboiler](https://github.com/volatiletech/sqlboiler).
